### PR TITLE
Finished Aurora load and tests : Caleb

### LINKED ISSUE #36: Create Quadhoral Aurorawatch load.py
As an amateur stargazer, I want the latest aurora information from aurorawatch API to be loaded into a database so that it can be displayed on a dashboard, so I can determine if it is worth 'auroragazing'.

- Use `psycopg2` to load data

### LINKED ISSUE #37: Test Quadhoral Aurorawatch load.py
As a developer, I want to test the Load.py script to ensure that the cleaned and transformed aurora data is correctly inserted into the database, so I can trust that the pipeline loads data accurately for future retrieval by users.

- Use `pytest` for this

### PR DESCRIPTION
**Pull request by CalebMathers**
This closes #37
This closes #36 

Wrote the load function for the AuroraWatch pipeline, along with tests.

### CODE REVIEW

File: aurorawatch-pipeline/requirements.txt [hunk 1]
```diff
@@ -1,4 +1,6 @@
 pylint
 pytest
 requests
-pytest-cov
\ No newline at end of file
+pytest-cov
+psycopg2-binary
+python-dotenv
```
Comment (olliejth) [inline aurorawatch-pipeline/requirements.txt:6]: Why did you get rid of pytest-cov?

File: aurorawatch-pipeline/aurora_load.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+'''Script to load the transformed data from the AuroraWatch UK API into the database.'''
+
+from os import environ as ENV
+import logging
+
+from psycopg2 import connect, extensions
+from dotenv import load_dotenv
+
+from aurora_extract import extract
+from aurora_transform import transform
```
Comment (CalebMathers) [inline aurorawatch-pipeline/aurora_load.py:10]: Nice separation of different import types, thanks.

File: aurorawatch-pipeline/aurora_load.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    return connect(f'''dbname={ENV["DB_NAME"]} user={ENV["DB_USER"]}
+                 host={ENV["DB_HOST"]} password={ENV["DB_PASSWORD"]} port={ENV["DB_PORT"]}''')
+
+
+def load_data(alert_data: tuple) -> None:
+    '''Loads the given AuroraWatch Tuple into the connected database.'''
+    query = '''INSERT INTO aurora_alert (alert_time, aurora_colour_id) VALUES (%s,%s);'''
+
+    with get_connection() as conn:
+        logging.info('Connected to database.')
```
Comment (CalebMathers) [inline aurorawatch-pipeline/aurora_load.py:26]: Good log to know if the connection was made or not

File: aurorawatch-pipeline/aurora_load.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+from aurora_transform import transform
+
+
+def get_connection() -> extensions.connection:
+    '''Returns a psycopg2 connection given the loaded environment variables.'''
+    load_dotenv()
+
+    return connect(f'''dbname={ENV["DB_NAME"]} user={ENV["DB_USER"]}
+                 host={ENV["DB_HOST"]} password={ENV["DB_PASSWORD"]} port={ENV["DB_PORT"]}''')
+
```
Comment (olliejth) [inline aurorawatch-pipeline/aurora_load.py:19]: DRY

Comment (olliejth) [reply aurorawatch-pipeline/aurora_load.py:19]: You could import this function from another file?

File: aurorawatch-pipeline/aurora_load.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            logging.info('New aurora alert data inserted.')
+
+
+if __name__ == "__main__":
+    logger = logging.getLogger(__name__)
+    logging.basicConfig(level=logging.INFO)
+
+    extracted_data = extract()
+    transformed_data = transform(extracted_data)
+    load_data(transformed_data)
```
Comment (olliejth) [inline aurorawatch-pipeline/aurora_load.py:39]: Very streamlined

File: aurorawatch-pipeline/test_aurora_load.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+'''Contains tests for the load script'''
+from unittest import mock
+from aurora_load import load_data
+
+
+class TestLoadData:
+    '''Tests for the load_data function'''
+    @mock.patch('aurora_load.get_connection')
```
Comment (olliejth) [inline aurorawatch-pipeline/test_aurora_load.py:8]: Could have a gap between these lines for consistency?

File: aurorawatch-pipeline/test_aurora_load.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        '''Tests that valid data doesn't cause an error'''
+        mock_conn = mock.MagicMock()
+        mock_cursor = mock.MagicMock()
+        mock_get_connection.return_value.__enter__.return_value = mock_conn
+        mock_conn.cursor.return_value.__enter__.return_value = mock_cursor
+
+        load_data(valid_insertion_data)
+        mock_conn.commit.assert_called_once()
+
+    @mock.patch('aurora_load.get_connection')
```
Comment (AlexOBarnes) [inline aurorawatch-pipeline/test_aurora_load.py:19]: Good idempotency

File: aurorawatch-pipeline/test_aurora_load.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
+'''Contains tests for the load script'''
+from unittest import mock
+from aurora_load import load_data
+
+
+class TestLoadData:
```
Comment (AlexOBarnes) [inline aurorawatch-pipeline/test_aurora_load.py:6]: Good use of classes

File: aurorawatch-pipeline/aurora_load.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+
+def load_data(alert_data: tuple) -> None:
+    '''Loads the given AuroraWatch Tuple into the connected database.'''
+    query = '''INSERT INTO aurora_alert (alert_time, aurora_colour_id) VALUES (%s,%s);'''
+
+    with get_connection() as conn:
+        logging.info('Connected to database.')
+        with conn.cursor() as cur:
+            cur.execute(query, alert_data)
```
Comment (AlexOBarnes) [inline aurorawatch-pipeline/aurora_load.py:28]: Have you considered using batch insert?

File: aurorawatch-pipeline/test_aurora_load.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    @mock.patch('aurora_load.get_connection')
+    def test_load_data_empty_list(self, mock_get_connection):
+        '''Tests that empty data doesn't result in an error'''
+        mock_conn = mock.MagicMock()
+        mock_cursor = mock.MagicMock()
+        mock_get_connection.return_value.__enter__.return_value = mock_conn
+        mock_conn.cursor.return_value.__enter__.return_value = mock_cursor
+
+        load_data(())
+        mock_conn.commit.assert_called_once()
```
Comment (olliejth) [inline aurorawatch-pipeline/test_aurora_load.py:28]: Nice but it looks like this should have been finished by lunch at least?

Comment (AlexOBarnes) [reply aurorawatch-pipeline/test_aurora_load.py:28]: I agree i agree i agree

**Review by AlexOBarnes — APPROVED** [review]

File: aurorawatch-pipeline/test_aurora_load.py (hunk 10 above)
Comment (olliejth) [reply aurorawatch-pipeline/test_aurora_load.py:28]: This guy gets it.

**Review by olliejth — APPROVED** [review]
Nice robust pipeline

**Pull request merged**